### PR TITLE
[WIP] Jenkins UI Themes

### PR DESCRIPTION
**Pull request by tfennelly**
I've evolved the [uithemes-plugin](https://github.com/jenkinsci/uithemes-plugin) a bit further in the last few weeks, to a stage now where (on this branch) a user can perform personalized (per login) theme configurations such as:
1. Status Balls/Orbs
2. Page Header
3. Icons
4. Console/terminal (ala #1272 from @kevinburke)

**[DOWNLOAD from CI Server](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/1875/artifact/war/target/jenkins.war)**

More info on [uithemes-plugin](https://github.com/jenkinsci/uithemes-plugin), as well as short video:

<a href="http://youtu.be/ZlD0zf1VCAs" target="_blank"></a>

Other possible themes:
1. Left side panel
2. Weather/health icons
3. ??

### CODE REVIEW

**Comment by olivergondza** [comment]
Is there any way to move most of it into plugin (and without bundling other plugins to core)?

**Comment by tfennelly** [comment]
The plugin is not a requirement - it's just needed if you want to allow users to configure it. If the plugin is not installed then users get all the defaults, which is the same styling as today.

As for "moving" these changes... most all of the changes in core for this are simply refactorings of stuff that was already there in .css files. There is the addition of some new theme implementations, but that's quite small and seem way more convenient in core, but yes, could be moved out.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -871,6 +872,9 @@ protected void doRun() throws Exception {
             if (LOG_STARTUP_PERFORMANCE)
                 LOGGER.info(String.format("Took %dms for complete Jenkins startup",
                         System.currentTimeMillis()-start));
+
+            // Install the core UI Themes
+            CoreThemeIntsaller.installCoreThemes();
```
Comment (olivergondza) [inline core/src/main/java/jenkins/model/Jenkins.java]: This should use [`@Initializer`](http://javadoc.jenkins-ci.org/hudson/init/Initializer.html) on `CoreThemeIntsaller.installCoreThemes`.

File: core/src/main/java/jenkins/uithemes/CoreThemeIntsaller.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.io.IOException;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class CoreThemeIntsaller {
```
Comment (olivergondza) [inline core/src/main/java/jenkins/uithemes/CoreThemeIntsaller.java]: typo

File: core/src/main/java/hudson/PluginManager.java [hunk 3]
```diff
@@ -506,6 +508,15 @@ protected boolean filter(Method e) {
...
         LOGGER.info("Plugin " + p.getShortName()+":"+p.getVersion() + " dynamically installed");
+
+        // If this plugin is a UIThemeContributor, tell the UIThemesProcessor about it.
+        // TODO: I don't see a plugin uninstall sequence ?
+        if (isUIThemesPluginInstalled()) {
+            Plugin pluginInstance = p.getPlugin();
+            if (pluginInstance instanceof UIThemeContributor) {
+                UIThemesProcessor.getInstance().addContributor((UIThemeContributor) pluginInstance);
+            }
+        }
```
Comment (olivergondza) [inline core/src/main/java/hudson/PluginManager.java]: I prefer introducing plugin installation/lifecycle listener uithemes would implement.

**Comment by olivergondza** [comment]
> The plugin is not a requirement - it's just needed if you want to allow users to configure it. If the plugin is not installed then users get all the defaults, which is the same styling as today.

But it is bundled - installed by default.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (tfennelly) [inline core/src/main/java/jenkins/model/Jenkins.java]: Ah cool thanks. I was not aware of that annotation. Thanks.

File: core/src/main/java/jenkins/uithemes/CoreThemeIntsaller.java (hunk 2 above)
Comment (tfennelly) [inline core/src/main/java/jenkins/uithemes/CoreThemeIntsaller.java]: I actually did look at that a few times and thought something "looked funny" about it :) Thanks.

File: core/src/main/java/hudson/PluginManager.java (hunk 3 above)
Comment (tfennelly) [inline core/src/main/java/hudson/PluginManager.java]: Yeah that would be nicer if it existed. I don't think it exists though, right? Might be overkill introducing it yet though.

**Comment by tfennelly** [comment]
@olivergondza It is installed, but can be uninstalled and styling will continue to work. I think it makes sense to have it bundled by default, but maybe that's my naivety.

**Comment by tfennelly** [comment]
This seems to have it's jocks in a bundle. Going to recreate the PR.

**Comment by tfennelly** [comment]
@olivergondza @daniel-beck I recreated this PR as #1524 .  Didn't need to recreate it in hindsight, but it's done now. Sorry about that.

The plugin is no long bundled by default and all the non "classic" theme stuff is now out in the uithemes-plugin. Not bundling it in Jenkins takes from the UX, but if it helps move things along then that's good.

**Pull request closed without merging**